File: content-solutions/graph-advanced/topo-course-order.py
import heapq


def solution(n, prereqs):
    graph = [[] for _ in range(n)]
    indeg = [0] * n
    for a, b in prereqs:
        graph[a].append(b)
        indeg[b] += 1
    heap = [v for v in range(n) if indeg[v] == 0]
    heapq.heapify(heap)
    order = []
    while heap:
        v = heapq.heappop(heap)
        order.append(v)
        for w in graph[v]:
            indeg[w] -= 1
            if indeg[w] == 0:
                heapq.heappush(heap, w)
    return order if len(order) == n else []

File: content-solutions/graph-advanced/topo-course-order.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] prereqs) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        int[] indeg = new int[n];
        for (int[] p : prereqs) {
            graph.get(p[0]).add(p[1]);
            indeg[p[1]]++;
        }
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for (int v = 0; v < n; v++) if (indeg[v] == 0) heap.offer(v);
        int[] order = new int[n];
        int count = 0;
        while (!heap.isEmpty()) {
            int v = heap.poll();
            order[count++] = v;
            for (int w : graph.get(v)) {
                indeg[w]--;
                if (indeg[w] == 0) heap.offer(w);
            }
        }
        return count == n ? order : new int[0];
    }
}
